### PR TITLE
Add Unitary calculation to qsim

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Adds the ability to compute the unitary matrix of a circuit to qsim in C++.
Next steps will be to link this up with op code and fuzz test, followed by implementing a layer that simplifies the workflow.

### CODE REVIEW

**Comment by zaqqwerty** [comment]
Could I get some context on the application of this machinery?  I assume the application is currently performance bottlenecked, since cirq circuits already provide the `unitary` method?

**Comment by MichaelBroughton** [comment]
Just like we have an op to compute the final state vectors of a circuit batch this would eventually be used in an op to compute the unitary matrices of a batch of circuits.

File: tensorflow_quantum/core/qsim/unitary_space_slow.cc [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    }
+  }
+}
+
+tensorflow::Status UnitarySpaceSlow::ApplyGate1(const float* matrix) {
+  float mat_00_r, mat_01_r, mat_10_r, mat_11_r;
+  float mat_00_i, mat_01_i, mat_10_i, mat_11_i;
+
+  auto state = GetRawUnitary();
+  mat_00_r = dot_one_r(state[0], state[1], state[4], state[5], 0, matrix);
```
Comment (zaqqwerty) [inline tensorflow_quantum/core/qsim/unitary_space_slow.cc:304]: Would be clearer to make the unitary be called `data` as above

File: tensorflow_quantum/core/qsim/unitary_space.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                          const float* matrix) = 0;
+
+  // Function to apply a one-qubit gate if there is one qubit in the Unitary.
+  // Implementations are given the option to return an error.
+  virtual tensorflow::Status ApplyGate1(const float* matrix) = 0;
+
+  // Set to identity matrix.
+  virtual void SetIdentity() = 0;
+
+  // Get the amplitude at the given state index.
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space.h]: Fixed.

File: tensorflow_quantum/core/qsim/unitary_space.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  virtual std::complex<float> GetEntry(const uint64_t i,
+                                       const uint64_t j) const = 0;
+
+  // Set the amplitude at the given state index
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space.h]: Fixed.

File: tensorflow_quantum/core/qsim/unitary_space.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  virtual void SetEntry(const uint64_t i, const uint64_t j,
+                        const std::complex<float>& val) = 0;
+
+ protected:
+  float* state_;
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space.h]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space_slow.cc [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    : UnitarySpace(num_qubits, num_threads) {}
+
+UnitarySpaceSlow::~UnitarySpaceSlow() { DeleteUnitary(); }
+
+UnitarySpaceType UnitarySpaceSlow::GetType() const {
+  return UnitarySpaceType::USLOW;
+}
+
+void UnitarySpaceSlow::CreateUnitary() {
+  state_ = (float*)malloc(sizeof(float) * size_);
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_slow.cc]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space_slow.cc [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+void UnitarySpaceSlow::DeleteUnitary() {
+  if (GetRawUnitary() != NULL) {
+    free(state_);
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_slow.cc]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space_slow.cc [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    state_ = NULL;
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_slow.cc]: Done.

File: tensorflow_quantum/core/qsim/unitary_space.cc [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  return tensorflow::Status::OK();
+}
+
+bool UnitarySpace::Valid() const {
+  // TODO: more roubust test?
+  return state_ != nullptr;
+}
+
+float* UnitarySpace::GetRawUnitary() const { return state_; };
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space.cc]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space.cc [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    CalcMatrix4(gate.qubits[0], gate.qubits[1], gate.gates, matrix);
+    ApplyGate2(gate.qubits[0], gate.qubits[1], matrix);
+  }
+
+  return tensorflow::Status::OK();
+}
+
+bool UnitarySpace::Valid() const {
+  // TODO: more roubust test?
+  return state_ != nullptr;
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space.cc]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space_slow.h [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                          const float* matrix) override;
+
+  // Function to apply a one-qubit gate if there is one qubit in the Unitary.
+  // Implementations are given the option to return an error.
+  virtual tensorflow::Status ApplyGate1(const float* matrix) override;
+
+  // Set to identity matrix.
+  virtual void SetIdentity() override;
+
+  // Get the amplitude at the given state index.
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_slow.h]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space_slow.h [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  virtual std::complex<float> GetEntry(const uint64_t i,
+                                       const uint64_t j) const override;
+
+  // Set the amplitude at the given state index
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_slow.h]: Changed.

File: tensorflow_quantum/core/qsim/unitary_space.h [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+// Contains the allowed Unitary labels
+enum UnitarySpaceType { USLOW };
+
+// Handles calculations of unitary matrices that a circuit enacts.
+class UnitarySpace {
+ public:
+  UnitarySpace(const uint64_t num_qubits, const uint64_t num_threads)
+      : state_(NULL),
+        size_(uint64_t{4} * (uint64_t{1} << num_qubits) *
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space.h]: Oops fixed.

File: tensorflow_quantum/core/qsim/unitary_space_test.cc [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+    }
+  }
+
+  Eigen::MatrixXcf ref_u =
+      Eigen::Map<Eigen::Matrix<std::complex<float>, Eigen::Dynamic,
+                               Eigen::Dynamic, Eigen::RowMajor>>(
+          (std::complex<float>*)unitary->GetRawUnitary(), 8, 8);
+
+  // clang-format off
+  float full_gate[] = {1.0,  2.0,  3.0,  4.0,  5.0,  6.0,  7.0,  8.0,  9.0,  10.0, 11.0,
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_test.cc:237]: Done.

File: tensorflow_quantum/core/qsim/unitary_space_test.cc [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+  for (int i = 0; i < 2; i++) {
+    for (int j = 0; j < 2; j++) {
+      EXPECT_EQ(unitary->GetEntry(i, j).real(), matrix_i[4 * i + 2 * j]);
+      EXPECT_EQ(unitary->GetEntry(i, j).imag(), matrix_i[4 * i + 2 * j + 1]);
+    }
+  }
+}
+
+TEST(UnitaryGate, UpdateSimpleTwoQubit) {
+  const std::array<float, 32> matrix_i = {
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_test.cc:287]: Added

File: tensorflow_quantum/core/qsim/unitary_space_slow.cc (hunk 1 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/qsim/unitary_space_slow.cc]: Done.

**Pull request merged**